rotkehlchen/db/history_events.py

```diff
@@ -401,35 +402,40 @@ def get_history_events(
         base_suffix = f'{HISTORY_BASE_ENTRY_FIELDS}, {EVM_EVENT_FIELDS}, {ETH_STAKING_EVENT_FIELDS} {ALL_EVENTS_DATA_JOIN}'  # noqa: E501
         bindings = []
         type_idx = 1 if group_by_event_ids is True else 0
-        special_free_query = False
-        free_query_group_by = maybe_filter_ignore_asset(filter_query)
+        final_clause = ''
 
         if has_premium is True:
+            filters, filters_bindings = filter_query.prepare(with_order=False, with_pagination=False)  # noqa: E501
+            bindings.extend(filters_bindings)
             if group_by_event_ids is True:
-                base_prefix += 'COUNT(*), '
-            base_query = f'{base_prefix} {base_suffix}'
+                base_query = (
+                    f'{base_prefix} COUNT(*), * FROM (SELECT {base_suffix} WHERE event_identifier '
+                    f'IN (SELECT DISTINCT event_identifier FROM history_events {filters})) '
+                )
+                if filter_query.group_by is not None:
+                    final_clause += f' {filter_query.group_by.prepare()}'
+            else:
+                base_query = f'{base_prefix} {base_suffix} {filters}'
         else:
+            filters, filters_bindings = filter_query.prepare(with_order=True, with_pagination=False)  # noqa: E501
             if group_by_event_ids is True:
-                special_free_query = True  # a bit ugly conditions to keep limit at groups for free users  # noqa: E501
-                free_query_group_by += 'GROUP BY event_identifier'
                 base_query = (
-                    f'{base_prefix} * FROM (SELECT COUNT(*), {base_suffix} '
-                    f'{free_query_group_by} ORDER BY timestamp DESC, sequence_index ASC LIMIT ?) '
+                    f'{base_prefix} COUNT(*), * FROM (SELECT {base_suffix} WHERE event_identifier '
+                    f'IN (SELECT DISTINCT event_identifier FROM history_events {filters} LIMIT ?))'
                 )
+                bindings.extend(filters_bindings + [self.free_history_events_limit])
+                if filter_query.group_by is not None:
+                    final_clause += f' {filter_query.group_by.prepare()}'
```

Comment:
in both branches (premium and notpremium) this seems to be the same logic. we already have in line 404 for `if group_by_event_ids` so you can do there

```python
if group_by_event_ids:
    type_idx = 
    if filter_query.group_by is not None:
        final_clause += f' {filter_query.group_by.prepare()}'
else:
    type_idx = 0